providers/websockets/binance/ws_data_handler.go

```diff
@@ -0,0 +1,161 @@
+		return resp, nil, nil
+	}
+
+	// Unmarshal the message as a stream message. If the message fails to be unmarshaled, this means
+	// that we received an unknown message type.
+	if err := json.Unmarshal(message, &streamMsg); err != nil {
+		return resp, nil, fmt.Errorf("failed to unmarshal message %w", err)
+	}
+
+	switch streamMsg.GetStreamType() {
+	case TickerStream:
+		// Ticker stream is sent every 1000ms and contains the latest price of a ticker.
+		var tickerResp TickerMessageResponse
+		if err := json.Unmarshal(message, &tickerResp); err != nil {
+			return resp, nil, fmt.Errorf("failed to unmarshal ticker message %w", err)
+		}
+
+		h.logger.Debug("received ticker message", zap.String("ticker", tickerResp.Data.Ticker))
+		resp, err := h.parsePriceUpdateMessage(tickerResp.Data.Ticker, tickerResp.Data.LastPrice)
+		return resp, nil, err
+	case AggregateTradeStream:
+		// Aggregate trade stream is sent when a trade is executed on the Binance exchange.
+		var aggTradeResp AggregatedTradeMessageResponse
+		if err := json.Unmarshal(message, &aggTradeResp); err != nil {
+			return resp, nil, fmt.Errorf("failed to unmarshal aggregate trade message %w", err)
+		}
+
+		h.logger.Debug("received aggregate trade message", zap.String("ticker", aggTradeResp.Data.Ticker))
+		resp, err := h.parsePriceUpdateMessage(aggTradeResp.Data.Ticker, aggTradeResp.Data.Price)
+		return resp, nil, err
+	default:
+		return resp, nil, fmt.Errorf("unknown stream type %s", streamMsg.Stream)
+	}
+}
+
+// CreateMessages is used to create a message to send to Binance. This is used to subscribe to
+// the given tickers. This is called when the connection to the data provider is first established.
+// Notably, the tickers have a unique identifier that is used to identify the messages going back
+// and forth. This unique identifier is the same one sent in the initial subscription.
+func (h *WebSocketHandler) CreateMessages(
+	tickers []types.ProviderTicker,
+) ([]handlers.WebsocketEncodedMessage, error) {
+	instruments := make([]string, 0)
+
+	for _, ticker := range tickers {
+		instruments = append(instruments, ticker.GetOffChainTicker())
+		h.cache.Add(ticker)
+	}
+
+	return h.NewSubscribeRequestMessage(instruments)
+}
+
+// HeartBeatMessages is not used for Binance. Heartbeats are handled on an ad-hoc basis when
+// messages are received from the Binance websocket API.
+//
+// ref: https://developers.binance.com/docs/binance-spot-api-docs/web-socket-streams#aggregate-trade-streams
+func (h *WebSocketHandler) HeartBeatMessages() ([]handlers.WebsocketEncodedMessage, error) {
+	return nil, nil
+}
+
+// Copy is used to create a copy of the WebSocketHandler.
+func (h *WebSocketHandler) Copy() types.PriceWebSocketDataHandler {
+	return &WebSocketHandler{
+		logger:     h.logger,
+		ws:         h.ws,
+		cache:      types.NewProviderTickers(),
+		messageIDs: make(map[int64][]string),
+	}
+}
```

Comment:
could you just call `NewWebSocketHandler`? might be safer if the constructor changes overtime, and one forgets to update this manual constructor here

Reply:
we might to deal with a panic if for whatever reason the `NewWebSocketHandler` returns an error. We've also just stuck to reconstructing the handler in the other providers this way so probably just prefer sticking to this for the time being.

Reply:
at some point we talked about creating a parent class for this exact reason, probably not high prio for the time being.